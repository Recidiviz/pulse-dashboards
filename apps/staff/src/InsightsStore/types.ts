// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { PromiseType } from "utility-types";

/**
 * Represents the internal state of a generator function annotated with `mobx.flow`.
 * Supports methods that only yield a single function call (or multiple that
 * all return the same type), since there is currently no way to represent
 * multiple yielded types in order at the type level.
 */
export type FlowMethod<
  FunctionToYield extends (...args: any[]) => Promise<unknown>,
  FlowReturn,
> = Generator<
  ReturnType<FunctionToYield>,
  FlowReturn,
  PromiseType<ReturnType<FunctionToYield>>
>;

export type StringMap2D<Value> = Map<string, Map<string, Value>>;
