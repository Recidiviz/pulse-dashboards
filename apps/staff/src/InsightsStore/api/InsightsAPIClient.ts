// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  ActionStrategy,
  actionStrategySchema,
  ClientEvent,
  clientEventSchema,
  ClientInfo,
  clientInfoSchema,
  ExcludedSupervisionOfficer,
  excludedSupervisionOfficerSchema,
  insightsConfigSchema,
  MetricBenchmark,
  metricBenchmarkSchema,
  SupervisionOfficer,
  supervisionOfficerFixture,
  SupervisionOfficerMetricEvent,
  supervisionOfficerMetricEventSchema,
  supervisionOfficerSchema,
  SupervisionOfficerSupervisor,
  supervisionOfficerSupervisorSchema,
  supervisionOfficerSupervisorsFixture,
  supervisionOfficerVitalsMetricFixture,
  SupervisionVitalsMetric,
  UserInfo,
  userInfoSchema,
} from "~datatypes";

import { formatDateToISO } from "../../utils";
import type { InsightsStore } from "../InsightsStore";
import {
  ActionStrategySurfacedEvent,
  InsightsAPI,
  PatchUserInfoProps,
} from "./interface";

export class InsightsAPIClient implements InsightsAPI {
  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly insightsStore: InsightsStore) {}

  private get apiStore() {
    return this.insightsStore.rootStore.apiStore;
  }

  private get baseUrl() {
    return `${import.meta.env.VITE_NEW_BACKEND_API_URL}/outliers/${this.tenantId}`;
  }

  get tenantId(): string {
    const { currentTenantId } = this.insightsStore.rootStore;
    if (!currentTenantId) {
      throw new Error(`Attempted to fetch data with undefined tenantId`);
    }
    return currentTenantId;
  }

  async init() {
    const endpoint = `${this.baseUrl}/configuration`;
    const fetchedData = await this.apiStore.get(endpoint);
    return insightsConfigSchema.parse(fetchedData.config);
  }

  async actionStrategies(supervisorPseudoId: string): Promise<ActionStrategy> {
    const endpoint = `${this.baseUrl}/action_strategies/${supervisorPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    return actionStrategySchema.parse(fetchedData);
  }

  async patchActionStrategies(
    props: ActionStrategySurfacedEvent,
  ): Promise<ActionStrategySurfacedEvent> {
    const endpoint = `${this.baseUrl}/action_strategies/${props.userPseudonymizedId}`;
    const fetchedData = await this.apiStore.patch(endpoint, props);
    const {
      userPseudonymizedId,
      officerPseudonymizedId,
      actionStrategy,
      timestamp,
    } = fetchedData;
    return {
      userPseudonymizedId,
      officerPseudonymizedId,
      actionStrategy,
      timestamp,
    };
  }

  async userInfo(userPseudoId: string): Promise<UserInfo> {
    const endpoint = `${this.baseUrl}/user-info/${userPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    return userInfoSchema.parse(fetchedData);
  }

  async patchUserInfo(
    userPseudoId: string,
    props: PatchUserInfoProps,
  ): Promise<UserInfo> {
    const endpoint = `${this.baseUrl}/user-info/${userPseudoId}`;
    const fetchedData = await this.apiStore.patch(endpoint, props);
    return userInfoSchema.parse(fetchedData);
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const endpoint = `${this.baseUrl}/benchmarks`;
    const fetchedData = await this.apiStore.get(endpoint);
    const benchmarkData = fetchedData.metrics as Array<unknown>;
    return benchmarkData.map((d) => {
      return metricBenchmarkSchema.parse(d);
    });
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const endpoint = `${this.baseUrl}/supervisors`;
    const fetchedData = await this.apiStore.get(endpoint);
    const supervisorData = fetchedData.supervisors as Array<unknown>;
    return supervisorData.map((b) =>
      supervisionOfficerSupervisorSchema.parse(b),
    );
  }

  async officersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionOfficer>> {
    const endpoint = `${this.baseUrl}/supervisor/${supervisorPseudoId}/officers`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officers as Array<unknown>;
    return officerData.map((b) => supervisionOfficerSchema.parse(b));
  }

  async excludedOfficersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<ExcludedSupervisionOfficer>> {
    const endpoint = `${this.baseUrl}/supervisor/${supervisorPseudoId}/excluded_officers`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officers as Array<unknown>;
    return officerData.map((b) => excludedSupervisionOfficerSchema.parse(b));
  }

  async supervisionOfficer(
    officerPseudoId: string,
  ): Promise<SupervisionOfficer> {
    const endpoint = `${this.baseUrl}/officer/${officerPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officer as unknown;
    return supervisionOfficerSchema.parse(officerData);
  }

  async excludedSupervisionOfficer(
    officerPseudoId: string,
  ): Promise<ExcludedSupervisionOfficer> {
    // TODO: (#6044) Convert "/excluded_officer/" to "/officer/" once all officers are served from the same endpoint.
    const endpoint = `${this.baseUrl}/excluded_officer/${officerPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const officerData = fetchedData.officer as unknown;
    return excludedSupervisionOfficerSchema.parse(officerData);
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string,
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const endpoint = `${this.baseUrl}/officer/${officerPseudoId}/events?metric_id=${metricId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const eventsData = fetchedData.events as Array<unknown>;
    return eventsData.map((b) => supervisionOfficerMetricEventSchema.parse(b));
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const endpoint = `${this.baseUrl}/client/${clientPseudoId}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const clientData = fetchedData.client as unknown;
    return clientInfoSchema.parse(clientData);
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date,
  ): Promise<Array<ClientEvent>> {
    const endpoint = `${
      this.baseUrl
    }/client/${clientPseudoId}/events?period_end_date=${formatDateToISO(
      endDate,
    )}`;
    const fetchedData = await this.apiStore.get(endpoint);
    const eventsData = fetchedData.events as Array<unknown>;
    return eventsData.map((b) => clientEventSchema.parse(b));
  }

  // TODO #6671 - Instead of using fixture data, make request to endpoint once it is deployed
  async vitalsForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionVitalsMetric>> {
    const supervisor = supervisionOfficerSupervisorsFixture.find(
      (s) => s.pseudonymizedId === supervisorPseudoId,
    );
    if (!supervisor) return [];
    return supervisionOfficerVitalsMetricFixture.map((metric) => {
      return {
        ...metric,
        vitalsMetrics: metric.vitalsMetrics.filter((metricForOfficer) => {
          const officersForSupervisor = supervisionOfficerFixture
            .filter((o) =>
              o.supervisorExternalIds.includes(supervisor?.externalId),
            )
            .map((o) => o.pseudonymizedId);
          return officersForSupervisor.includes(
            metricForOfficer.officerPseudonymizedId,
          );
        }),
      };
    });
  }

  async vitalsForOfficer(
    officerPseudoId: string,
  ): Promise<Array<SupervisionVitalsMetric>> {
    return supervisionOfficerVitalsMetricFixture.map((metric) => {
      return {
        ...metric,
        vitalsMetrics: metric.vitalsMetrics.filter(
          (metricForOfficer) =>
            officerPseudoId === metricForOfficer.officerPseudonymizedId,
        ),
      };
    });
  }
}
