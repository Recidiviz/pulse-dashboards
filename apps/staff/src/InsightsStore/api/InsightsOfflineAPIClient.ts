/* eslint-disable class-methods-use-this */
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { differenceInDays, subDays } from "date-fns";

import type { InsightsStore } from "../InsightsStore";
import { ClientEvent } from "../models/ClientEvent";
import { ClientInfo } from "../models/ClientInfo";
import { MetricBenchmark } from "../models/MetricBenchmark";
import { LATEST_END_DATE } from "../models/offlineFixtures/constants";
import { leadershipUserInfoFixture } from "../models/offlineFixtures/UserInfoFixture";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { SupervisionOfficerMetricEvent } from "../models/SupervisionOfficerMetricEvent";
import { SupervisionOfficerSupervisor } from "../models/SupervisionOfficerSupervisor";
import { UserInfo } from "../models/UserInfo";
import { InsightsAPI, PatchUserInfoProps } from "./interface";

export class InsightsOfflineAPIClient implements InsightsAPI {
  private pseudoIdToEditableUserInfo: Map<
    string,
    { metadata: PatchUserInfoProps }
  > = new Map();

  // eslint-disable-next-line no-useless-constructor
  constructor(public readonly insightsStore: InsightsStore) {}

  async init() {
    const config = (
      await import("../models/offlineFixtures/InsightsConfigFixture")
    ).InsightsConfigFixture;

    return config;
  }

  async userInfo(userPseudoId: string): Promise<UserInfo> {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    const matchingSupervisor = supervisionOfficerSupervisorsFixture.find(
      (supervisor) => supervisor.pseudonymizedId === userPseudoId,
    );

    if (matchingSupervisor) {
      return {
        role: "supervision_officer_supervisor",
        entity: matchingSupervisor,
        ...(this.pseudoIdToEditableUserInfo.get(userPseudoId) ?? {
          metadata: {
            hasSeenOnboarding: false,
          },
        }),
      };
    }

    return {
      ...leadershipUserInfoFixture,
      ...(this.pseudoIdToEditableUserInfo.get(userPseudoId) ?? {
        metadata: {
          hasSeenOnboarding: false,
        },
      }),
    };
  }

  async patchUserInfo(
    userPseudoId: string,
    props: PatchUserInfoProps,
  ): Promise<UserInfo> {
    this.pseudoIdToEditableUserInfo.set(userPseudoId, { metadata: props });
    return this.userInfo(userPseudoId);
  }

  async metricBenchmarks(): Promise<MetricBenchmark[]> {
    const { metricBenchmarksFixture } = await import(
      "../models/offlineFixtures/MetricBenchmarkFixture"
    );
    return metricBenchmarksFixture;
  }

  async supervisionOfficerSupervisors(): Promise<
    SupervisionOfficerSupervisor[]
  > {
    const { supervisionOfficerSupervisorsFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerSupervisor"
    );
    return supervisionOfficerSupervisorsFixture;
  }

  async officersForSupervisor(
    supervisorPseudoId: string,
  ): Promise<Array<SupervisionOfficer>> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    return supervisionOfficerFixture.filter(
      (o) => `hashed-${o.supervisorExternalId}` === supervisorPseudoId,
    );
  }

  async supervisionOfficer(
    officerPseudoId: string,
  ): Promise<SupervisionOfficer> {
    const { supervisionOfficerFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerFixture"
    );

    const officerFixture = supervisionOfficerFixture.find(
      (o) => o.pseudonymizedId === officerPseudoId,
    );

    if (!officerFixture)
      throw new Error(`Officer ${officerPseudoId} not present in fixture data`);

    return officerFixture;
  }

  async supervisionOfficerMetricEvents(
    officerPseudoId: string,
    metricId: string,
  ): Promise<SupervisionOfficerMetricEvent[]> {
    const { supervisionOfficerMetricEventFixture } = await import(
      "../models/offlineFixtures/SupervisionOfficerMetricEventFixture"
    );
    return supervisionOfficerMetricEventFixture.filter(
      (e) => e.metricId === metricId,
    );
  }

  async clientInfo(clientPseudoId: string): Promise<ClientInfo> {
    const { clientInfoFixture } = await import(
      "../models/offlineFixtures/ClientInfoFixture"
    );

    const clientInfo = clientInfoFixture[clientPseudoId];

    if (!clientInfo) {
      throw new Error(`Client ${clientPseudoId} not present in fixture data`);
    }

    return clientInfo;
  }

  async clientEvents(
    clientPseudoId: string,
    endDate: Date,
  ): Promise<Array<ClientEvent>> {
    const { clientEventFixture } = await import(
      "../models/offlineFixtures/ClientEventFixture"
    );

    // adjust fixture dates relative to the input
    const dateOffsetInDays = differenceInDays(LATEST_END_DATE, endDate);

    return clientEventFixture.map((event) => ({
      ...event,
      eventDate: subDays(event.eventDate, dateOffsetInDays),
    }));
  }
}
