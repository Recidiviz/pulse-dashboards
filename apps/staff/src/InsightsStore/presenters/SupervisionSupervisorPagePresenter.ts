// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { flowResult, makeAutoObservable } from "mobx";

import { SupervisionOfficer, SupervisionOfficerSupervisor } from "~datatypes";
import {
  Hydratable,
  HydratesFromSource,
  HydrationState,
} from "~hydration-utils";

import { InsightsSupervisionStore } from "../stores/InsightsSupervisionStore";
import { HighlightedOfficersDetail } from "./types";
import { getHighlightedOfficersByMetric } from "./utils";

export class SupervisionSupervisorPagePresenter implements Hydratable {
  protected hydrator: HydratesFromSource;

  constructor(
    protected supervisionStore: InsightsSupervisionStore,
    public supervisorPseudoId: string,
  ) {
    makeAutoObservable(this, {}, { autoBind: true });
    this.hydrator = new HydratesFromSource({
      populate: async () => {
        await Promise.all([...this.populateMethods()]);
      },
      expectPopulated: this.expectPopulated(),
    });
  }

  get isWorkflowsEnabled() {
    const { userStore } = this.supervisionStore.insightsStore.rootStore;

    // Check if...
    return (
      // ...the user has allowed navigation to workflows and...
      userStore.getRoutePermission("workflowsSupervision") &&
      // ...if the active feature variant for supervisorHomepageWorkflows is enabled.
      !!userStore.activeFeatureVariants.supervisorHomepageWorkflows
    );
  }

  get isVitalsEnabled() {
    return this.supervisionStore.isVitalsEnabled;
  }

  /**
   * Returns an array of promises representing the methods required to populate
   * the necessary data for this presenter.
   */
  populateMethods() {
    return [
      flowResult(
        this.supervisionStore.populateOfficersForSupervisor(
          this.supervisorPseudoId,
        ),
      ),
      flowResult(this.supervisionStore.populateSupervisionOfficerSupervisors()),
    ];
  }

  /**
   * Returns an array of expectations for whether the necessary data has been populated.
   */
  expectPopulated() {
    return [this.expectOfficersPopulated, this.expectSupervisorPopulated];
  }

  get pageTitle() {
    return `${this.supervisorInfo?.displayName} Overview`;
  }

  get labels() {
    return this.supervisionStore.labels;
  }

  /**
   * Provides a string with the current time period.
   * @returns The time period string or `undefined` if not available.
   */
  get timePeriod(): string | undefined {
    return this.supervisionStore.benchmarksTimePeriod;
  }

  /**
   * Checks if the current user has access to all supervisors.
   * @returns `true` if the user can access all supervisors, `false` otherwise.
   */
  get userCanAccessAllSupervisors() {
    return this.supervisionStore.userCanAccessAllSupervisors;
  }

  /**
   * Provides a list of all officers excluded from outcomes in this supervisor's unit.
   * @returns An array of `SupervisionOfficer` or `undefined` if data is not available.
   */
  get officersIncludedInOutcomes(): SupervisionOfficer[] | undefined {
    return this.supervisionStore.officersBySupervisorPseudoId
      .get(this.supervisorPseudoId)
      ?.filter((o) => o.includeInOutcomes === true);
  }

  /**
   * Combines and returns all officers, both included and excluded, under this supervisor.
   * @returns An array of `SupervisionOfficer` (empty if data not available).
   */
  get allOfficers(): SupervisionOfficer[] {
    return (
      this.supervisionStore.officersBySupervisorPseudoId.get(
        this.supervisorPseudoId,
      ) ?? []
    );
  }

  /**
   * Provides a list of all officers excluded from outcomes in this supervisor's unit.
   * @returns An array of `SupervisionOfficer` or `undefined` if data is not available.
   */
  get officersExcludedFromOutcomes(): SupervisionOfficer[] | undefined {
    return this.supervisionStore.officersBySupervisorPseudoId
      .get(this.supervisorPseudoId)
      ?.filter((o) => o.includeInOutcomes !== true);
  }

  /**
   * Returns officers of this supervisor in the top X percent of officers in the state,
   * grouped by metric.
   * @returns An array of objects containing the metric, top X percent criteria, and info about
   * officers meeting the top X percent criteria.
   */
  get highlightedOfficersByMetric(): HighlightedOfficersDetail[] {
    return getHighlightedOfficersByMetric(
      this.supervisionStore.metricConfigsById,
      this.officersIncludedInOutcomes,
      this.supervisionStore.officersOutcomesBySupervisorPseudoId.get(
        this.supervisorPseudoId,
      ),
    );
  }

  /**
   * Provides information about the currently selected supervisor.
   * @returns The supervisor record, or `undefined` if not yet fetched.
   */
  get supervisorInfo(): SupervisionOfficerSupervisor | undefined {
    return this.supervisionStore.supervisorInfo(this.supervisorPseudoId);
  }

  /**
   * Provides information about the currently selected supervisor's
   * supervision location, whether it is by unit or by district
   */
  get supervisionLocationInfo(): {
    locationLabel: string;
    supervisionLocationForListPage?: string | null;
    supervisionLocationForSupervisorPage?: string | null;
  } {
    return this.supervisionStore.supervisionLocationInfo(
      this.supervisorPseudoId,
    );
  }

  /**
   * Tracks the event when a supervisor's page is viewed by the user.
   */
  trackViewed(): void {
    const { userPseudoId } =
      this.supervisionStore.insightsStore.rootStore.userStore;
    this.supervisionStore.insightsStore.rootStore.analyticsStore.trackInsightsSupervisorPageViewed(
      {
        supervisorPseudonymizedId: this.supervisorPseudoId,
        viewedBy: userPseudoId,
      },
    );
  }

  private expectOfficersPopulated() {
    if (
      !this.supervisionStore.officersBySupervisorPseudoId.has(
        this.supervisorPseudoId,
      )
    )
      throw new Error("failed to populate officers with outliers");
  }

  /**
   * Asserts that supervisor data has been populated.
   * @throws An error if supervisor data is not populated.
   */
  private expectSupervisorPopulated() {
    if (!this.supervisorInfo) throw new Error("failed to populate supervisor");
  }

  /**
   * Initiates hydration for all data needed within this presenter class
   */
  async hydrate(): Promise<void> {
    return this.hydrator.hydrate();
  }

  get hydrationState(): HydrationState {
    return this.hydrator.hydrationState;
  }
}
