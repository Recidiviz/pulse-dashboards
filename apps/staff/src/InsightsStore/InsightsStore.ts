// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { makeAutoObservable, reaction, runInAction } from "mobx";

import { isDemoMode, isOfflineMode, isTestEnv } from "~client-env-utils";

import { RootStore } from "../RootStore";
import { InsightsAPIClient } from "./api/InsightsAPIClient";
import { InsightsOfflineAPIClient } from "./api/InsightsOfflineAPIClient";
import { InsightsAPI } from "./api/interface";
import { InsightsSupervisionStore } from "./stores/InsightsSupervisionStore";
import { FlowMethod } from "./types";

export class InsightsStore {
  supervisionStore?: InsightsSupervisionStore;

  apiClient: InsightsAPI;

  constructor(public rootStore: RootStore) {
    this.apiClient = this.getApiClient();

    makeAutoObservable(this);

    // reset the store for each new tenant and if the current user changes (such as via impersonation)
    reaction(
      () => {
        return {
          tenant: this.rootStore.currentTenantId,
          user: this.rootStore.userStore.user,
        };
      },
      () => {
        runInAction(() => this.reset());
      },
    );
  }

  reset() {
    this.supervisionStore = undefined;
    // TODO #5198 Do not reset apiClient once CA uses real InsightsAPIClient
    this.apiClient = this.getApiClient();
  }

  getApiClient(): InsightsAPI {
    return isOfflineMode() ||
      isTestEnv() ||
      isDemoMode() ||
      // TODO #5198 Remove this check once CA uses real InsightsAPIClient
      this.rootStore.currentTenantId === "US_CA"
      ? new InsightsOfflineAPIClient(this)
      : new InsightsAPIClient(this);
  }

  /**
   * Creates this.supervisionStore for the current tenant, if it does not already exist.
   * This is a MobX flow method and should be called with mobx.flowResult.
   */
  *populateSupervisionStore(): FlowMethod<InsightsAPI["init"], void> {
    if (this.supervisionStore) return;

    const config = yield this.apiClient.init();

    this.supervisionStore = new InsightsSupervisionStore(this, config);
  }
}
