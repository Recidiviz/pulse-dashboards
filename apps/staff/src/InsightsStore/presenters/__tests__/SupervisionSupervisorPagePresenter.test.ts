// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2025 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { configure } from "mobx";

import {
  InsightsConfigFixture,
  supervisionOfficerSupervisorsFixture,
} from "~datatypes";
import { unpackAggregatedErrors } from "~hydration-utils";

import { RootStore } from "../../../RootStore";
import UserStore from "../../../RootStore/UserStore";
import { InsightsOfflineAPIClient } from "../../api/InsightsOfflineAPIClient";
import { InsightsStore } from "../../InsightsStore";
import { InsightsSupervisionStore } from "../../stores/InsightsSupervisionStore";
import { SupervisionSupervisorPagePresenter } from "../SupervisionSupervisorPagePresenter";

const testSupervisor = supervisionOfficerSupervisorsFixture[0];

let store: InsightsSupervisionStore;
let presenter: SupervisionSupervisorPagePresenter;
let rootStore: RootStore;

beforeEach(async () => {
  configure({ safeDescriptors: false });

  // USER STORE =========================================================
  vi.spyOn(UserStore.prototype, "userPseudoId", "get").mockImplementation(
    () => testSupervisor.pseudonymizedId,
  );

  vi.spyOn(UserStore.prototype, "isRecidivizUser", "get").mockImplementation(
    () => false,
  );

  vi.spyOn(UserStore.prototype, "stateCode", "get").mockImplementation(
    () => "US_ID",
  );

  // SUPERVISION STORE =================================================
  store = new InsightsSupervisionStore(
    new InsightsStore(new RootStore()),
    InsightsConfigFixture,
  );

  vi.spyOn(store, "userCanAccessAllSupervisors", "get").mockReturnValue(true);
  rootStore = store.insightsStore.rootStore;

  rootStore.workflowsRootStore.opportunityConfigurationStore.mockHydrated();

  presenter = new SupervisionSupervisorPagePresenter(
    store,
    testSupervisor.pseudonymizedId,
  );
});

afterEach(() => {
  vi.restoreAllMocks();
  configure({ safeDescriptors: true });
});

test("supervisorInfo", async () => {
  await presenter.hydrate();

  const { supervisorInfo } = presenter;
  expect(supervisorInfo).toMatchSnapshot();
});

test("supervisorLocationInfo", async () => {
  await presenter.hydrate();

  const { supervisionLocationInfo } = presenter;
  expect(supervisionLocationInfo).toMatchSnapshot();
});

test("returns the officers without outcomes data", async () => {
  await presenter.hydrate();

  expect(presenter.officersExcludedFromOutcomes).toMatchSnapshot();
});

test("returns the officers with outcomes data", async () => {
  await presenter.hydrate();

  expect(presenter.officersIncludedInOutcomes).toMatchSnapshot();
});

test("supervisorId not found in supervisionOfficerSupervisors", async () => {
  presenter = new SupervisionSupervisorPagePresenter(store, "nonExistentId");
  await presenter.hydrate();
  expect(presenter.hydrationState).toMatchInlineSnapshot(`
    {
      "error": [AggregateError: Expected data failed to populate],
      "status": "failed",
    }
  `);
  expect(unpackAggregatedErrors(presenter)).toMatchInlineSnapshot(`
    [
      [Error: failed to populate supervisor],
    ]
  `);
});

test("supervisorId not found in officersBySupervisor", async () => {
  vi.spyOn(store, "officersBySupervisorPseudoId", "get").mockImplementation(
    () => new Map(),
  );

  await presenter.hydrate();

  expect(
    store.officersBySupervisorPseudoId.has(testSupervisor.pseudonymizedId),
  ).toBeFalse();
  expect(presenter.hydrationState).toMatchInlineSnapshot(`
    {
      "error": [AggregateError: Expected data failed to populate],
      "status": "failed",
    }
  `);

  expect(unpackAggregatedErrors(presenter)).toMatchInlineSnapshot(`
    [
      [Error: failed to populate officers with outliers],
    ]
  `);
});

test("supervisor has no officer outcomes", async () => {
  vi.spyOn(
    InsightsOfflineAPIClient.prototype,
    "outcomesForSupervisor",
  ).mockResolvedValue([]);

  await presenter.hydrate();

  expect(presenter.hydrationState).toMatchInlineSnapshot(`
    {
      "status": "hydrated",
    }
  `);
});
