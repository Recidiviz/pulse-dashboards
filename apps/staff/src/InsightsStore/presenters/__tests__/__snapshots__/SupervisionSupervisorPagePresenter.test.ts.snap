// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns the officers with outcomes data 1`] = `
[
  {
    "avgDailyPopulation": 54,
    "displayName": "Walter Harris",
    "district": "Unit 1",
    "email": "officer1@recidiviz.org",
    "externalId": "OFFICER4",
    "fullName": {
      "givenNames": "Walter",
      "surname": "Harris",
    },
    "includeInOutcomes": true,
    "pseudonymizedId": "hashed-so1",
    "supervisorExternalIds": [
      "agonzalez123",
      "dlee456",
    ],
    "zeroGrantOpportunities": [
      "LSU",
      "pastFTRD",
    ],
  },
  {
    "avgDailyPopulation": 12,
    "displayName": "Jack Hernandez",
    "district": "Unit 1",
    "email": null,
    "externalId": "so2",
    "fullName": {
      "givenNames": "Jack",
      "surname": "Hernandez",
    },
    "includeInOutcomes": true,
    "pseudonymizedId": "hashed-so2",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
  {
    "avgDailyPopulation": 32,
    "displayName": "Jason Nelson",
    "district": "Unit 1",
    "email": "officer3@recidiviz.org",
    "externalId": "so3",
    "fullName": {
      "givenNames": "Jason",
      "surname": "Nelson",
    },
    "includeInOutcomes": true,
    "pseudonymizedId": "hashed-so3",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
  {
    "avgDailyPopulation": 43,
    "displayName": "Harriet Boyd",
    "district": "Unit 1",
    "email": "officer9@recidiviz.org",
    "externalId": "so9",
    "fullName": {
      "givenNames": "Harriet",
      "surname": "Boyd",
    },
    "includeInOutcomes": true,
    "pseudonymizedId": "hashed-so9",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
  {
    "avgDailyPopulation": 43,
    "displayName": "Stephen Mann",
    "district": "Unit 1",
    "email": "officer10@recidiviz.org",
    "externalId": "so10",
    "fullName": {
      "givenNames": "Stephen",
      "surname": "Mann",
    },
    "includeInOutcomes": true,
    "pseudonymizedId": "hashed-so10",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
]
`;

exports[`returns the officers without outcomes data 1`] = `
[
  {
    "avgDailyPopulation": null,
    "displayName": "John Harris",
    "district": "Unit 3",
    "email": "officer6@recidiviz.org",
    "externalId": "so6",
    "fullName": {
      "givenNames": "John",
      "surname": "Harris",
    },
    "pseudonymizedId": "hashed-so6",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
  {
    "avgDailyPopulation": 8,
    "displayName": "Larry Hernandez",
    "district": "Unit 2",
    "email": "officer7@recidiviz.org",
    "externalId": "so7",
    "fullName": {
      "givenNames": "Larry",
      "surname": "Hernandez",
    },
    "pseudonymizedId": "hashed-so7",
    "supervisorExternalIds": [
      "agonzalez123",
    ],
  },
]
`;

exports[`supervisorInfo 1`] = `
{
  "displayName": "Alejandro D Gonzalez",
  "email": "mock-email",
  "externalId": "agonzalez123",
  "fullName": {
    "givenNames": "Alejandro",
    "middleNames": "D",
    "surname": "Gonzalez",
  },
  "hasOutliers": true,
  "pseudonymizedId": "hashed-agonzalez123",
  "supervisionLocationForListPage": "Region 1",
  "supervisionLocationForSupervisorPage": "Unit 1",
}
`;

exports[`supervisorLocationInfo 1`] = `
{
  "locationLabel": "unit",
  "supervisionLocationForListPage": "Region 1",
  "supervisionLocationForSupervisorPage": "1",
}
`;
